Detectors/PHOS/calib/PHOSCalibWorkflow/src/PHOSEnergyCalibrator.cxx

```diff
@@ -0,0 +1,233 @@
+    EventHeader h = {0};
+    h.mMarker = 16383;
+    h.mBC = tr.getBCData().bc;
+    mDigits.push_back(h.mDataWord);
+    mDigits.push_back(tr.getBCData().orbit);
+
+    int iclu = 0;
+    int firstCluInEvent = tr.getFirstEntry();
+    int lastCluInEvent = firstCluInEvent + tr.getNumberOfObjects();
+
+    mBuffer->startNewEvent(); // mark stored clusters to be used for Mixing
+    for (int i = firstCluInEvent; i < lastCluInEvent; i++) {
+      const FullCluster& clu = clusters[i];
+
+      fillTimeMassHisto(clu);
+      // bool isGood = checkCluster(clu);
+
+      auto cluList = clu.getElementList();
+      for (auto ce = cluList->begin(); ce != cluList->end(); ce++) {
+        short absId = ce->absId;
+        //Fill cells from cluster for next iterations
+        short adcCounts = ce->energy / mCalibParams->getGain(absId);
+        // Need to chale LG gain too to fit dynamic range
+        if (!ce->isHG) {
+          adcCounts /= mCalibParams->getHGLGRatio(absId);
+        }
+        CalibDigit d = {0};
+        d.mAddress = absId;
+        d.mAdcAmp = adcCounts;
+        d.mHgLg = ce->isHG;
+        d.mCluster = (i - firstCluInEvent) % kMaxCluInEvent;
+        mDigits.push_back(d.mDataWord);
+        if (i - firstCluInEvent > kMaxCluInEvent) {
+          //Normally this is not critical as indexes are used "locally", i.e. are compared to previous/next
+          LOG(INFO) << "Too many clusters per event:" << i - firstCluInEvent << ", apply more strict selection; clusters with same indexes will appear";
+        }
+      }
+    }
+  }
+}
+void PHOSEnergySlot::clear()
+{
+  mHistos.reset();
+  mDigits.clear();
+}
+
+void PHOSEnergySlot::fillTimeMassHisto(const FullCluster& clu)
+{
+  // Fill time distributions only for cells in cluster
+  auto cluList = clu.getElementList();
+  for (auto ce = cluList->begin(); ce != cluList->end(); ce++) {
+    short absId = ce->absId;
+    if (ce->isHG) {
+      if (ce->energy > mEminHGTime) {
+        mHistos.fill(ETCalibHistos::kTimeHGPerCell, absId, ce->time);
+      }
+      mHistos.fill(ETCalibHistos::kTimeHGSlewing, ce->time, ce->energy);
+    } else {
+      if (ce->energy > mEminLGTime) {
+        mHistos.fill(ETCalibHistos::kTimeLGPerCell, absId, ce->time);
+      }
+      mHistos.fill(ETCalibHistos::kTimeLGSlewing, ce->time, ce->energy);
+    }
+  }
+
+  //Real and Mixed inv mass distributions
+  // prepare TLorentsVector
+  float posX, posZ;
+  clu.getLocalPosition(posX, posZ);
+  TVector3 vec3;
+  mGeom->local2Global(clu.module(), posX, posZ, vec3);
+  vec3 -= mVertex;
+  float e = clu.getEnergy();
+  short absId;
+  mGeom->relPosToAbsId(clu.module(), posX, posZ, absId);
+
+  vec3 *= 1. / vec3.Mag();
+  TLorentzVector v(vec3.X() * e, vec3.Y() * e, vec3.Z() * e, e);
+  // Fill calibration histograms for all cells, even bad, but partners in inv, mass should be good
+  bool isGood = checkCluster(clu);
+  for (short ip = mBuffer->size(); ip--;) {
+    const TLorentzVector& vp = mBuffer->getEntry(ip);
```

Comment:
actually I see that you don't need copy, please change getEntry to return the reference.